R/tostring.R

```diff
@@ -1,3 +1,247 @@
+## state completely sucks and I hate it, but
+## we need a pdf device open to calculate the
+## print width of strings, and we can't be opening
+## a new one every time we want to
+font_dev_state <- new.env()
+font_dev_state$open <- FALSE
+font_dev_state$fontspec <- list()
+font_dev_state$spacewidth <- NA_real_
+font_dev_state$ismonospace <- NA
+font_dev_state$max_ratio <- NA_real_
+font_dev_state$dev_num <- NA_integer_
+
+
+cwidth_inches_unsafe <- function(x) {
+  convertWidth(unit(1, "strwidth", x), "inches", valueOnly = TRUE)
+}
+
+## returns whether it opened a new device or not
+#' Activate font state
+#'
+#' @param fontspec (`font_spec`)\cr a font_spec object specifying the font information to use for
+#'   calculating string widths and heights, as returned by [font_spec()].
+#' @param silent (`logical(1)`)\cr If `FALSE`, the default, a warning will be
+#'   emitted if this function switches away from an active graphics device.
+#'
+#' @details The font device state is an environment with
+#' four variables guaranteed to be set:
+#'
+#' \describe{
+#'   \item{`open`}{(`logical(1)`)\cr whether a device is already open with font info}
+#'   \item{`fontspec`}{(`font_spec`)\cr the font specification, if any, that is currently active (`list()` if none is).}
+#'   \item{`spacewidth`}{(`numeric(1)`)\cr the width of the space character in the currently active font.}
+#'   \item{`ismonospace`}{(`logical(1)`)\cr whether the specified font is monospaced.}
+#' }
+#'
+#' `open_font_dev` opens a pdf device with the specified font
+#' only if there is not one currently open with the same font.
+#' If a new device is opened, it caches `spacewidth` and
+#' `ismonospace` for use in `nchar_ttype`).
+#'
+#' `close_font_dev` closes any open font state device
+#' and clears the cached values.
+#'
+#' @return
+#' - `open_font_dev` returns a logical value indicating whether a *new* pdf device was opened.
+#' - `close_font_dev` returns `NULL`.
+#'
+#' In both cases the value is returned invisibly.
+#'
+#' @examples
+#' open_font_dev(font_spec("Times"))
+#' nchar_ttype("Hiya there", font_spec("Times"))
+#' close_font_dev()
+#'
+#' @export
+open_font_dev <- function(fontspec, silent = FALSE) {
+  ## XXX to remove for debugging only
```

Comment:
should we keep it only for debugging or remove it? if only debugging I would keep internal, It is always accessible with ::: or by doing devtools::load_all()

Reply:
Sorry, my comment may hae been unclear, `open_font_dev` is an important exported api entrypoint, the comment is about the lines directly below it, which yes can probably be removed, the real "For maintainers debuggin switch" is at line 77 where you remove the FALSE. Will push a commit to remove that comment

Reply:
pushed

Reply:
understood!!! thanks